joint_state_publisher/joint_state_publisher/joint_state_publisher.py

```diff
@@ -59,6 +59,51 @@ def _init_joint(self, minval, maxval, zeroval):
             joint['effort'] = 0.0
         return joint
 
+    def init_sdf(self, robot):
+        robot = robot.getElementsByTagName('model')[0]
+        # Find all non-fixed joints
+        for child in robot.childNodes:
+            if child.nodeType is child.TEXT_NODE:
+                continue
+            if child.localName != 'joint':
+                continue
+            jtype = child.getAttribute('type')
+            if jtype in ('gearbox', 'revolute2', 'ball', 'screw', 'universal', 'fixed'):
+                continue
+            name = child.getAttribute('name')
+            self.joint_list.append(name)
```

Comment:
I think this line should be moved below, right before we do `self.free_joints[name] = joint`.

The main reason is that there are additional checks down below, so if a joint violates one of the limits, then we could get into a state where it was added to `self.joint_list`, but not `self.free_joints`.  To avoid that, we shouldn't change any object state until after we have done all of our checks.